x/src/lint/engine/runner.rs

```diff
@@ -0,0 +1,183 @@
+
+        let project_ctx = ProjectContext::new(self.config.project_root);
+
+        if let Some(content_linters) = self.config.content_linters {
+            let file_list = self.get_file_list()?;
+
+            // TODO: This should probably be a worker queue with a thread pool or something.
+
+            let file_ctxs: Vec<_> = file_list
+                .iter()
+                .map(|path| FileContext::new(project_ctx, path))
+                .collect();
+
+            'outer: for file_ctx in file_ctxs {
+                let linters_to_run = content_linters
+                    .iter()
+                    .copied()
+                    .filter_map(|linter| match linter.pre_run(&file_ctx) {
+                        Ok(RunStatus::Executed) => Some(Ok(linter)),
+                        Ok(RunStatus::Skipped(reason)) => {
+                            let source = file_ctx.source(linter.name());
+                            skipped.push((source, reason));
+                            None
+                        }
+                        Err(err) => Some(Err(err)),
+                    })
+                    .collect::<Result<Vec<_>>>()?;
+
+                if linters_to_run.is_empty() {
+                    // No linters to run for this file -- no point loading it.
+                    continue;
+                }
+
+                // Load up the content for this file.
+                let content_ctx = file_ctx.load()?;
+
+                for linter in linters_to_run {
+                    let source = content_ctx.source(linter.name());
+                    let mut formatter = LintFormatter::new(source, &mut messages);
+
+                    match linter.run(&content_ctx, &mut formatter)? {
+                        RunStatus::Executed => {
+                            // Yay! Lint ran successfully.
+                        }
+                        RunStatus::Skipped(reason) => {
+                            skipped.push((source, reason));
+                        }
+                    }
+
+                    if self.config.fail_fast && !messages.is_empty() {
+                        // At least one issue was found.
+                        break 'outer;
+                    }
+                }
+            }
+        }
+
+        Ok(LintResults { skipped, messages })
+    }
+
+    // ---
+    // Helper methods
+    // ---
+
+    fn get_file_list(&self) -> Result<Vec<&Path>> {
+        // TODO: abstract out SCM and command-running functionality.
+        let stdout = self.files_stdout()?;
+
+        Ok(stdout
+            .split(|b| b == &0)
+            .filter_map(|s| {
+                // TODO: make global exclusions configurable.
+                if !s.is_empty() && !s.starts_with(b"testsuite/libra-fuzzer/artifacts/") {
+                    // Well, this only works on Unix for now. Would be cool to get it working on
+                    // Windows at some point too, but it needs to be done with some care. See
```

Comment:
Given we have the raw bytes isn't using OsStr enough, or is the fact that git is probably outputting stuff incorrectly the source of the problem?

Reply:
The problem is that `OsStr::from_bytes` is only available on Unixes. I'll expand the comment to clarify that.

Reply:
Ah got it, I see now.